### PR TITLE
planner, expression: fix TableFullScan caused by extraHandleId not correctly found

### LINKED ISSUE #45036: TableRangeScan is not selected when autocommit=OFF
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
CREATE TABLE ads_txn (
  `cusno` varchar(10) NOT NULL,
  `txn_dt` varchar(8) NOT NULL,
  `unn_trno` decimal(22,0) NOT NULL,
  `aml_cntpr_accno` varchar(64) DEFAULT NULL,
  `acpayr_accno` varchar(35) DEFAULT NULL,
  PRIMARY KEY (`cusno`,`txn_dt`,`unn_trno`) NONCLUSTERED
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
PARTITION BY LIST COLUMNS(`txn_dt`)
(PARTITION `p20000101` VALUES IN ('20000101'),
 PARTITION `p20220101` VALUES IN ('20220101'),
 PARTITION `p20230516` VALUES IN ('20230516'));

analyze table ads_txn;

explain update ads_txn s set aml_cntpr_accno = trim(acpayr_accno) where s._tidb_rowid between 1 and 100000;
set autocommit=OFF;
explain update ads_txn s set aml_cntpr_accno = trim(acpayr_accno) where s._tidb_rowid between 1 and 100000;

```

### 2. What did you expect to see? (Required)
Both plans are TableRangeScan:
```
+--------------------------+----------+-----------+---------------+--------------------------------------------------+
| id                       | estRows  | task      | access object | operator info                                    |
+--------------------------+----------+-----------+---------------+--------------------------------------------------+
| Update_4                 | N/A      | root      |               | N/A                                              |
| └─TableReader_7          | 10000.00 | root      | partition:all | data:TableRangeScan_6                            |
|   └─TableRangeScan_6     | 10000.00 | cop[tikv] | table:s       | range:[1,100000], keep order:false, stats:pseudo |
+--------------------------+----------+-----------+---------------+--------------------------------------------------+
3 rows in set (0.00 sec)
```

### 3. What did you see instead (Required)
TableFullScan when autocommit=OFF:
```
+-------------------------------+----------+-----------+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------+
| id                            | estRows  | task      | access object | operator info                                                                                                                                     |
+-------------------------------+----------+-----------+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------+
| Update_5                      | N/A      | root      |               | N/A                                                                                                                                               |
| └─Projection_6                | 8000.00  | root      |               | test.ads_txn.cusno, test.ads_txn.txn_dt, test.ads_txn.unn_trno, test.ads_txn.aml_cntpr_accno, test.ads_txn.acpayr_accno, test.ads_txn._tidb_rowid |
|   └─SelectLock_7              | 8000.00  | root      |               | for update 0                                                                                                                                      |
|     └─TableReader_10          | 8000.00  | root      | partition:all | data:Selection_9                                                                                                                                  |
|       └─Selection_9           | 8000.00  | cop[tikv] |               | ge(test.ads_txn._tidb_rowid, 1), le(test.ads_txn._tidb_rowid, 100000)                                                                             |
|         └─TableFullScan_8     | 10000.00 | cop[tikv] | table:s       | keep order:false, stats:pseudo                                                                                                                    |
+-------------------------------+----------+-----------+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------+
6 rows in set (0.00 sec)
```

### 4. What is your TiDB version? (Required)
v6.5.2

### PR DESCRIPTION
**Pull request by chrysan**
### What problem does this PR solve?

Issue Number: close #45036

Problem Summary:
Fix the wrong logic to get extraHandleId. And this part needs further refactoring.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by hawkingrei — APPROVED** [review]

File: planner/core/integration_test.go [hunk 1]
```diff
@@ -5332,3 +5332,30 @@ func TestIssue45033(t *testing.T) {
 				                        from   t3 alias3) alias4
 				                where  alias4.c2 = alias2.alias_col1);`).Check(testkit.Rows("0"))
 }
+
+// https://github.com/pingcap/tidb/issues/45036
+func TestIssue45036(t *testing.T) {
```
Comment (hawkingrei) [inline planner/core/integration_test.go:5337]: If this test cannot be categorized specifically, it can be placed under ```planner/core/issuetest```.

Comment (chrysan) [reply planner/core/integration_test.go]: Done

**Review by qw4990 — APPROVED** [review]

File: planner/core/issuetest/planner_issue_test.go [hunk 2]
```diff
@@ -118,3 +118,30 @@ func TestIssue42732(t *testing.T) {
...
+		"  `unn_trno` decimal(22,0) NOT NULL,\n" +
+		"  `aml_cntpr_accno` varchar(64) DEFAULT NULL,\n" +
+		"  `acpayr_accno` varchar(35) DEFAULT NULL,\n" +
+		"  PRIMARY KEY (`cusno`,`txn_dt`,`unn_trno`) NONCLUSTERED\n" +
+		") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin\n" +
+		"PARTITION BY LIST COLUMNS(`txn_dt`)\n" +
+		"(PARTITION `p20000101` VALUES IN ('20000101'),\n" +
+		"PARTITION `p20220101` VALUES IN ('20220101'),\n" +
+		"PARTITION `p20230516` VALUES IN ('20230516'))")
+	tk.MustExec("analyze table ads_txn")
```
Comment (time-and-fate) [inline planner/core/issuetest/planner_issue_test.go:139]: Is this `analyze` unnecessary?

File: expression/schema.go [hunk 3]
```diff
@@ -237,6 +238,17 @@ func (s *Schema) MemoryUsage() (sum int64) {
...
 }
 
+// GetExtraHandleColumn gets the extra handle column.
+func (s *Schema) GetExtraHandleColumn() *Column {
+	columnLen := len(s.Columns)
+	if columnLen > 0 && s.Columns[columnLen-1].ID == model.ExtraHandleID {
+		return s.Columns[columnLen-1]
+	} else if columnLen > 1 && s.Columns[columnLen-2].ID == model.ExtraHandleID {
+		return s.Columns[columnLen-2]
+	}
```
Comment (time-and-fate) [inline expression/schema.go:248]: I think it's better to find from the last to the first column to defend against any future changes.

**Review by hawkingrei — APPROVED** [review]

File: planner/core/issuetest/planner_issue_test.go (hunk 2 above)
Comment (chrysan) [reply planner/core/issuetest/planner_issue_test.go:139]: Yes, to prepare global stats for dynamic prune.

File: expression/schema.go (hunk 3 above)
Comment (chrysan) [reply expression/schema.go:248]: I thought about it, but there could be performance downgrade if it's a long Columns slice and no handleID is found.

**Pull request merged**